### PR TITLE
ENH: Adds GNSS Receiver sensor

### PR DESCRIPTION
**Pull request by MateusStano**
## Pull request type

- [x] Code changes (bugfix, features)

## Checklist

- [x] Tests for the changes have been added (if needed)
- [x] Docs have been reviewed and added / updated
- [x] Lint (`black rocketpy/ tests/`) has passed locally 
- [x] All tests (`pytest tests -m slow --runslow`) have passed locally
- [x] `CHANGELOG.md` has been updated (if relevant)

## Changes

- Add class `GNSS`
- Merges develop and fixes conflicts with new implementations (mainly pylint and FlightPhases refactor)

The noise parameters of the `GNSS` are much simpler than from the other sensors. I made it so because it seems that errors are taken into account as `accuracy` rather then noise densities in GNSS sensors.

I am not sure if the way its done is the most robust way of doing this, but it was what I gathered from research

## Breaking change

- [ ] Yes
- [x] No

### CODE REVIEW

File: rocketpy/__init__.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
     Tail,
     TrapezoidalFins,
 )
-from .sensors import Accelerometer, Barometer, Gyroscope
+from .sensors import GNSS, Accelerometer, Barometer, Gyroscope
```
Comment (Gui-FernandesBR) [inline rocketpy/__init__.py:40]: I suggest `GNSSReceiver`, `GNSSUnit` or `GNSSSensor`.

Just `GNSS` doesn't make so much sense.

File: rocketpy/rocket/parachute.py [hunk 2]
```diff
@@ -190,6 +190,7 @@ def __evaluate_trigger_function(self, trigger):
         """This is used to set the triggerfunc attribute that will be used to
         interact with the Flight class.
         """
+        # pylint: disable=unused-argument disable=function-redefined
```
Comment (Gui-FernandesBR) [inline rocketpy/rocket/parachute.py:193]:
```suggestion
        # pylint: disable=unused-argument, function-redefined
```

File: rocketpy/sensors/barometer.py [hunk 3]
```diff
@@ -150,16 +150,12 @@ def measure(self, time, **kwargs):
...
             - relative_position : np.array
                 Position of the sensor relative to the rocket center of mass.
-            - gravity : float
-                Gravitational acceleration in m/s^2.
-            - pressure : Function
-                Atmospheric pressure profile as a function of altitude in Pa.
-            - elevation : float
-                Elevation of the launch site in meters.
+            - environment : Environment
+                Environment object containing the atmospheric conditions.
```
Comment (Gui-FernandesBR) [inline rocketpy/sensors/barometer.py:154]: Receiving an Environment? At a first glance, this seems odd to me. Let me give a try.

File: rocketpy/sensors/gnss.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    ):
+        """Initialize the GNSS sensor.
+
+        Parameters
+        ----------
+        sampling_rate : float
+            Sample rate of the sensor in Hz.
+        position_accuracy : float
+            Accuracy of the sensor interpreted as the standard deviation of the
+            position in meters. Default is 0.
```
Comment (Gui-FernandesBR) [inline rocketpy/sensors/gnss.py:50]: These instruments usually have different accuracy in different axes. I would expect an accuracy for z, another for x, and third one for y. And to be more accurate, we usually see errors being expressed in terms of longitude/latitude.

```suggestion
            Accuracy of the sensor interpreted as the standard deviation of the
            position in meters. Default is 0.
```

File: rocketpy/sensors/gnss.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        longitude = math.degrees(
+            lon1
+            + math.atan2(
+                math.sin(math.radians(bearing))
+                * math.sin(drift / earth_radius)
+                * math.cos(lat1),
+                math.cos(drift / earth_radius)
+                - math.sin(lat1) * math.sin(math.radians(latitude)),
+            )
+        )
```
Comment (Gui-FernandesBR) [inline rocketpy/sensors/gnss.py:120]: The haversine equation is being used in many parts of our code already. I think it's time for us to define it in `rocketpy.tools` and simply import it wherever it's necessary.

File: rocketpy/simulation/flight.py [hunk 6]
```diff
@@ -3173,6 +3175,32 @@ class attributes which are instances of the Function class. Usage
             encoding="utf-8",
         )
 
+    def export_sensor_data(self, file_name, sensor=None):
```
Comment (Gui-FernandesBR) [inline rocketpy/simulation/flight.py:3178]: It would be nice to also allow receiving the sensor's name

File: rocketpy/simulation/flight.py [hunk 7]
```diff
@@ -3173,6 +3175,32 @@ class attributes which are instances of the Function class. Usage
...
+        if sensor is None:
+            data_dict = {}
+            for used_sensor, measured_data in self.sensor_data.items():
+                data_dict[used_sensor.name] = measured_data
+        else:
+            # export data of only that sensor
+            data_dict = {}
+            data_dict[sensor.name] = self.sensor_data[sensor]
+        with open(file_name, "w") as file:
+            json.dump(data_dict, file)
```
Comment (Gui-FernandesBR) [inline rocketpy/simulation/flight.py:3201]: I wonder how is the .csv output like.

File: rocketpy/tools.py [hunk 8]
```diff
@@ -928,7 +928,7 @@ def quaternions_to_nutation(e1, e2):
     return (180 / np.pi) * 2 * np.arcsin(-((e1**2 + e2**2) ** 0.5))
 
 
-def euler_to_quaternions(yaw, pitch, roll):
+def euler_to_quaternions(roll, pitch, yaw):
```
Comment (Gui-FernandesBR) [inline rocketpy/tools.py:931]: Please also invert it in the docstring!

```suggestion
def euler_to_quaternions(roll, pitch, yaw): 
```

File: tests/unit/test_flight.py [hunk 9]
```diff
@@ -1,3 +1,5 @@
+import json
+import os
```
Comment (Gui-FernandesBR) [inline tests/unit/test_flight.py:2]: do you really use these modules (json and os) here?

**Review by Gui-FernandesBR — COMMENTED** [review]
First look review.
I already like what I see. Good work!

I would like to take a few minutes during the weekend to test it locally. 
But I'm ok if someone else reviews it first.
If that's the case, don't wait for my review.

Please fix the errors pointed out by `pylint`, `isort` and `flake8`

**Comment by Gui-FernandesBR** [comment]
This PR will probably conflict with #632 in the test files

File: rocketpy/sensors/gnss.py (hunk 5 above)
Comment (MateusStano) [reply rocketpy/sensors/gnss.py:120]: There is already a `haversine` function in `tools.py` but it does not seem to do what I need here. Lets leave this for another PR

File: rocketpy/simulation/flight.py (hunk 6 above)
Comment (MateusStano) [reply rocketpy/simulation/flight.py:3178]: Receive the sensor name to choose the sensor you want to export data from? 

You mean to receive the sensor name in the `sensor` argument or create another argument called `sensor_name`?

Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight.py:3178]: all the sensors objects have a "name" attribute, no?

File: rocketpy/sensors/gnss.py (hunk 4 above)
Comment (Gui-FernandesBR) [reply rocketpy/sensors/gnss.py:50]: > These instruments usually have different accuracy in different axes. I would expect an accuracy for z, another for x, and third one for y. And to be more accurate, we usually see errors being expressed in terms of longitude/latitude.

Actually, I would expect one "horizontal accuracy" and another "vertical accuracy".
This is what I'm currently familiar with. But my knowledge about specific GNSS instruments for rockets is almost zero.

All in all you can let only just a single accuracy here, it's okay. If someone else complains, we can get back to it in the future.


Example:
https://globalgis.lk/storage/2022/05/E800_Datasheet_20201111_ENG-H.pdf
image

File: rocketpy/sensors/gnss.py (hunk 5 above)
Comment (Gui-FernandesBR) [reply rocketpy/sensors/gnss.py:120]: there're `haversine` and `inverted_haversine` functions.
What you looking for is the `inverted_haversine` one.

Comment (MateusStano) [reply rocketpy/sensors/gnss.py]: https://github.com/RocketPy-Team/RocketPy/pull/636/commits/a28e7bd3170076d4b4923baf9a0bdb006e1d3192

File: rocketpy/simulation/flight.py (hunk 6 above)
Comment (MateusStano) [reply rocketpy/simulation/flight.py:3178]: https://github.com/RocketPy-Team/RocketPy/pull/636/commits/9ce718eaf809dc5b3059e05f08a313388af3a89e

File: rocketpy/__init__.py (hunk 1 above)
Comment (Gui-FernandesBR) [reply rocketpy/__init__.py]: [015f5ea](https://github.com/RocketPy-Team/RocketPy/pull/636/commits/015f5ea8e6191dca662d472b09728d340f32a839)

**Review by Gui-FernandesBR — APPROVED** [review]
LGTM

**Pull request merged**